### PR TITLE
WIP: StringVector

### PR DESCRIPTION
**Pull request by shashi**
StringVector is an array of strings that keeps strings in a single buffer and returns `WeakRefString` when indexed. This depends on Andreas's fork of WeakRefStrings.

Some features:

1. Fast serialization (see `s/stringvec` branch on JuliaDB)
2. Can mmap
4. supports `setindex!`, `push!` and `resize!`
3. `setindex!` with its own elements is fast: hence can `sort!` or `permute!` are fast and don't grow the buffer.

Drawbacks:

1. Doesn't GC old buffer regions when new strings are repeatedly set at the same index. We hope that this use case is rare (since `sort!` and `push!` are already optimized), transformations should use `map` to create new vectors.

### CODE REVIEW

File: src/csv.jl [hunk 1]
```diff
@@ -70,12 +70,12 @@ csvread(file::String, delim=','; kwargs...) = _csvread_f(file, delim; kwargs...)
 
 function csvread(file::IOStream, delim=','; kwargs...)
     mmap_data = Mmap.mmap(file)
-    _csvread(String(mmap_data), delim; kwargs...)
+    _csvread(WeakRefString(pointer(mmap_data), length(mmap_data)), delim; kwargs...)
```
Comment (shashi) [inline src/csv.jl]: @andreasnoack Are you sure we want this back?

File: src/util.jl [hunk 2]
```diff
@@ -157,13 +157,14 @@ end
...
+using WeakRefStrings
 using PooledArrays
 
+_pointer(x::WeakRefString, n) = x.ptr + n - 1
 _pointer(x::String, n) = pointer(x, n)
 
 @inline function nonallocating_setindex!(pa::PooledArray{T}, i, rng::StrRange, str::AbstractString) where {T}
-    # wstr = WeakRefString(_pointer(str, 1+rng.offset), rng.length)
-    wstr = str[(1:rng.length) + rng.offset] # so now it is allocating
+    wstr = WeakRefString(_pointer(str, 1+rng.offset), rng.length)
```
Comment (shashi) [inline src/util.jl]: Nice to have this back :-)

**Comment by davidanthoff** [comment]
Would this mean that by default one would get an array of ``WeakRefString``s now? And they are unsafe because they don't keep the memory region alive, right? If so, could we put this behind a flag so that one can opt into ``WeakRefString``s? It doesn't seem like a good default for me, and it would also (I think) more or less break the use-case in the Queryverse area.

Another question: I thought that on 0.7 small structs have been optimized to be stack allocated, even if they hold a reference to something on the heap. If that is so, couldn't one just use a ``SubString`` here?

File: src/csv.jl (hunk 1 above)
Comment (andreasnoack) [reply src/csv.jl]: No. I'm not. I just reverted the old commit. I remember timing this one and it didn't really matter so lets just keep `String`.

**Comment by andreasnoack** [comment]
To follow up on @davidanthoff's comment. I think we should keep this as a branch for a while to figure out the best way to do this. I'm also not too comfortable about leaking `WeakRefString`s to the surroundings. The speedup inside JuliaDB is so extreme that something like this is necessary though but it would be great to see if we can encapsulate it somehow. As David mentions, it might also be that we can implement this slightly differently in a more safe way in 0.7. We'll only know once we have migrated.

**Comment by shashi** [comment]
I agree that UnsafeStrings are a bad thing to expose to users!

I have put it behind an`unsafestrings` flag. Also I have moved the code to https://github.com/JuliaComputing/StringArrays.jl

I'm also considering having `getindex` return a `String` but `_getindex` returning `UnsafeString`. IndexedTables can use the `_getindex` function internally. What are your thoughts @davidanthoff ? Will you be able to use that easily in Query?

**Comment by shashi** [comment]
The approach I went with is parameterizing StringArray with the right eltype rather than defining different `getindex` functions. Converting between eltypes is cheap.

**Comment by shashi** [comment]
Here I've enabled StringArray use by default. The user will see an array of eltype `String` by default. IndexedTables will convert it to `WeakRefString{UInt8}` to do sortperm etc.

**Pull request merged**